sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/subquery.scala

```diff
@@ -83,6 +116,20 @@ object SubqueryExpression {
       case _ => false
     }.isDefined
   }
+
+  /**
+   * Clean the outer references by normalizing them to BindReference in the same way
+   * we clean up the arguments during LogicalPlan.sameResult. This enables to compare two
+   * plans which has subquery expressions. This method returns a [[CanonicalizedSubqueryExpr]]
+   * which wraps the underlying [[SubqueryExpression]].
+   */
+  def canonicalize(e: SubqueryExpression, attrs: AttributeSeq): CanonicalizedSubqueryExpr = {
+    // Normalize the outer references in the subquery plan.
+    val subPlan = e.plan.transformAllExpressions {
+      case OuterReference(r) => BindReferences.bindReference(r, attrs, allowFailures = true)
```

Comment:
@cloud-fan The local attributes would be bound when we call sameResult on the subquery plan ? We certainly can bind them here and if you don't mind, i wanted to explore this in a follow up in an effort to start considering the sub-plan in the hashCode computation (marked as a todo now in comment) as i wanted more testing on that. Please let me know what you think.